src/hotspot/cpu/ppc/stubGenerator_ppc.cpp

```diff
@@ -3544,6 +3544,442 @@ class StubGenerator: public StubCodeGenerator {
+    __ lxv(offsetLUT, 0, const_ptr);
+    __ load_const_optimized(const_ptr, (address)&maskLUT_URL_val, tmp_reg);
+    __ lxv(maskLUT, 0, const_ptr);
+    __ xxspltib(vec_special_case_char->to_vsr(), '_');
+    __ xxspltib(vec_special_case_offset, (unsigned char)US);
+    __ b(calculate_size);
+
+    // isURL = 0 (false)
+    __ bind(not_URL);
+    __ load_const_optimized(const_ptr, (address)&offsetLUT_val, tmp_reg);
+    __ lxv(offsetLUT, 0, const_ptr);
+    __ load_const_optimized(const_ptr, (address)&maskLUT_val, tmp_reg);
+    __ lxv(maskLUT, 0, const_ptr);
+    __ xxspltib(vec_special_case_char->to_vsr(), '/');
+    __ xxspltib(vec_special_case_offset, (unsigned char)SLS);
+
+    __ bind(calculate_size);
+
+    // Don't handle the last 4 characters of the source, because this
+    // VSX-based algorithm doesn't handle padding characters.  Also the
+    // vector code will always write 16 bytes of decoded data on each pass,
+    // but only the first 12 of those 16 bytes are valid data (16 base64
+    // characters become 12 bytes of binary data), so for this reason we
+    // need to subtract an additional 8 bytes from the source length, in
+    // order not to write past the end of the destination buffer.  The
+    // result of this subtraction implies that the non-instrinsic routine
+    // will be used to process the last 12 characters.
+    __ sub(sl, sl, sp);
+    __ subi(sl, sl, 12);
+
+    // Round sl down to the nearest multiple of block_size
+    __ clrrdi(sl, sl, block_size_clear);
+
+    // out starts at d + dp
+    __ add(out, d, dp);
+
+    // in starts at s + sp
+    __ add(in, s, sp);
+
+    // Address of the last byte of the source is (in + sl - 1)
+    __ add(end, in, sl);
+    __ subi(end, end, 1);
+
+    __ bind(unrolled_loop_start);
+
+    __ cmpd(CCR0, end, in);
+    __ blt_predict_not_taken(CCR0, unrolled_loop_exit);
+    for (unsigned unroll_cnt=0; unroll_cnt < loop_unrolls; unroll_cnt++) {
+        // We can use a static displacement in the load since it's always a
+        // multiple of 16, which is a requirement of lxv/stxv.  This saves
+        // an addi instruction.
+        __ lxv(input->to_vsr(), unroll_cnt * 16, in);
+        //
+        // Lookup
+        //
+        // Isolate the upper 4 bits of each character by shifting it right 4 bits
+        __ vsrb(higher_nibble, input, vec_4s);
+        // Isolate the lower 4 bits by masking
+        __ xxland(lower_nibble, input->to_vsr(), vec_0xfs);
+
+        // Get the offset (the value to subtract from the byte) by using
+        // a lookup table indexed by the upper 4 bits of the character
+        __ xxperm(offsets->to_vsr(), offsetLUT, higher_nibble->to_vsr());
+
+        // Find out which elemets are the special case character (isURL ? '/' : '-')
+        __ vcmpequb_(eq_special_case_char, input, vec_special_case_char);
+        //
+        // There's a (63/64)^16 = 77.7% chance that there are no special
```

Comment:
I think that assumes uniformly randomized data, is that a good assumption?   Is it measurably faster to skip around the xxsel instead of doing it unconditionally?

Reply:
Thanks for this question. I also stumbled over it when reviewing. I guess a branch which gets mispredicted in ~22% of the cases leads to a big performance loss. (In addition, the branch target is not aligned.)

Reply:
Yes, it assumes uniformly random data, but also recall that the unencoded data bytes get shifted by 2, 4, 6 bits into the encoded bytes, which I'm guessing would tend to make the data somewhat more uniform, even if the source data has low entropy.

That said, I didn't actually benchmark it.  I will do that to make sure there is a gain, and if there isn't I will remove the conditional branch.

Reply:
> I took a look at the VSX algo. I haven't looked much beyond it. I had a few questions I've inlined. It does look like a faithful VSX implementation of the linked algo.

I neglected to thank you for reviewing this code!  I realize there's quite a time commitment required to review this properly, and because of that I was having difficulty finding a second reviewer for the PPC64 portion.

Just to set expectations, I will be on vacation next week, so further commits won't be posted until the following week, but I will address all of your great feedback.  Thanks again!

Reply:
I just got done running a benchmark without the branch around the xxsel, and your hunch was right.  There's about a 9% performance gain in the benchmark with that branch dropped.  I also changed the previous instruction not to set the condition code, but I doubt that affected performance.  Both regression tests for Base64 encoding/decoding still pass.

The next set of commits will contain this change.